### PR TITLE
Verifiable presentation API for ION attestor

### LINKED ISSUE #59: Implement issuing a verifiable presentation
Include functionality for transforming from a VC to a VP. This process relates to the [W3C VC API](https://w3c-ccg.github.io/vc-api/#prove-presentation). See work in [didkit](https://github.com/spruceid/didkit/tree/main/cli#didkit-vc-issue-presentation) for reference.

### PR DESCRIPTION
**Pull request by edchapman88**
Closes #59.

### CODE REVIEW

File: trustchain-api/src/api.rs [hunk 1]
```diff
@@ -119,5 +128,230 @@ pub trait TrustchainVCAPI {
     }
 }
 
+#[async_trait]
+pub trait TrustchainVPAPI {
+    /// As a holder issue a verifiable presentation.
```
Comment (sgreenbury) [inline trustchain-api/src/api.rs:133]:
```suggestion
    /// Signs a presentation constructing a verifiable presentation.
```

File: trustchain-ion/src/attestor.rs [hunk 2]
```diff
@@ -168,11 +169,51 @@ impl Issuer for IONAttestor {
...
+        // Get the signing key.
+        let signing_key = self.signing_key(key_id)?;
+
+        let mut vp = presentation.clone();
+        // Check holder field is correctly populated
+        match presentation.holder.as_ref() {
+            Some(URI::String(holder)) => {
+                if holder != &self.did {
+                    vp.holder = Some(URI::String(self.did.clone()))
+                }
```
Comment (sgreenbury) [inline trustchain-ion/src/attestor.rs:196]: Following discussion, update to return an error variant related to incompatible holders.

File: trustchain-api/src/api.rs [hunk 3]
```diff
@@ -119,5 +128,230 @@ pub trait TrustchainVCAPI {
...
+        verifier: &IONVerifier<T>,
+    ) -> Result<(), PresentationError> {
+        // Verify contained credentials
+        let credentials = presentation
+            .verifiable_credential
+            .as_ref()
+            .ok_or(PresentationError::NoCredentialsPresent)?;
+
+        // https://gendignoux.com/blog/2021/04/01/rust-async-streams-futures-part1.html#unordered-buffering-1
+        // https://docs.rs/futures-util/latest/futures_util/stream/trait.TryStreamExt.html#method.try_for_each_concurrent
```
Comment (sgreenbury) [inline trustchain-api/src/api.rs:161]: To remove

File: trustchain-api/src/api.rs [hunk 4]
```diff
@@ -119,5 +128,230 @@ pub trait TrustchainVCAPI {
...
+        endpoint: &str,
+        ldp_options: Option<LinkedDataProofOptions>,
+    ) -> Result<Presentation, PresentationError> {
+        let resolver = get_ion_resolver(endpoint);
+        let attestor = IONAttestor::new(did);
+        Ok(attestor
+            .sign_presentation(&presentation, key_id, &resolver, ldp_options)
+            .await?)
+    }
+    /// Verifies a verifiable presentation. Analogous with [didkit](https://docs.rs/didkit/latest/didkit/c/fn.didkit_vc_verify_presentation.html).
```
Comment (sgreenbury) [inline trustchain-api/src/api.rs:147]:
```suggestion
    /// Verifies a verifiable presentation.
```

File: trustchain-api/src/api.rs [hunk 5]
```diff
@@ -119,5 +128,230 @@ pub trait TrustchainVCAPI {
...
+                                credential,
+                                ldp_options,
+                                root_event_time,
+                                verifier,
+                            )
+                            .await
+                            .map(|_| ())
+                            .map_err(|err| err.into());
+                            println!("done {}:  {}", idx, now_ms());
+                            v
```
Comment (sgreenbury) [inline trustchain-api/src/api.rs:187]: Remove assignment and print statements

File: trustchain-core/src/holder.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    }
+}
+
+impl From<KeyManagerError> for HolderError {
+    fn from(err: KeyManagerError) -> Self {
+        HolderError::KeyManager(err)
+    }
+}
+
+/// A holder signs a credential to generate a verifiable credential.
```
Comment (sgreenbury) [inline trustchain-core/src/holder.rs:32]:
```suggestion
/// A holder signs a presentation to generate a verifiable presentation.
```

File: trustchain-core/src/holder.rs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+use async_trait::async_trait;
+use ssi::did_resolve::DIDResolver;
+use ssi::vc::{LinkedDataProofOptions, Presentation};
+use thiserror::Error;
+
+/// An error relating to a Trustchain holder.
+#[derive(Error, Debug)]
+pub enum HolderError {
+    /// Wrapped error for SSI error.
+    #[error("A wrapped variant for an SSI error.")]
```
Comment (sgreenbury) [inline trustchain-core/src/holder.rs:13]: Fix display to include error data. Also fix this in `issuer.rs`.

File: trustchain-core/src/holder.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+pub trait Holder: Subject {
+    /// Attests to a given presentation of one or many credentials returning the presentation with a
+    /// proof. The `@context` of the presentation has linked-data fields strictly checked as part of
+    /// proof generation.
+    async fn sign_presentation<T: DIDResolver>(
+        &self,
+        presentation: &Presentation,
+        key_id: Option<&str>,
+        resolver: &T,
+        ldp_options: Option<LinkedDataProofOptions>,
```
Comment (sgreenbury) [inline trustchain-core/src/holder.rs:43]: Move (and rename to `linked_data_proof_options`) to the argument after `presentation` to have closer API to `Issuer` trait.

File: trustchain-core/src/utils.rs [hunk 9]
```diff
@@ -20,6 +20,16 @@ pub fn init() {
         // initialization code here
         let tempdir = tempfile::tempdir().unwrap();
         std::env::set_var(TRUSTCHAIN_DATA, Path::new(tempdir.as_ref().as_os_str()));
+
+        // TODO: write the required key_manager path in TRUSTCHAIN_DATA with: root, root-plus-1,
```
Comment (sgreenbury) [inline trustchain-core/src/utils.rs:24]: TODO: implement this

File: trustchain-core/src/vp.rs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+impl From<VerifierError> for PresentationError {
+    fn from(err: VerifierError) -> Self {
+        PresentationError::VerifierError(err)
+    }
+}
+
+#[cfg(test)]
+mod tests {}
```
Comment (sgreenbury) [inline trustchain-core/src/vp.rs:52]:
```suggestion
```

File: trustchain-ion/src/attestor.rs [hunk 11]
```diff
@@ -268,6 +309,49 @@ mod tests {
...
+        // assert holder field is not initially populated
+        assert!(presentation.holder.is_none());
+
+        // Attest to vp:
+        // .sign_presenatation now checks and sets the holder field of the presentation
+        // This has implications for the proof generation which is handled by the ssi library:
+        //   - ssi::ldp::ensure_or_pick_verification_relationship calls presentation.get_issuer()
+        //      which returns the holder (if Some)
+        //   - ensure_or_pick_verification_relationship tries to resolve the did and check it's
+        //      verification methods
```
Comment (sgreenbury) [inline trustchain-ion/src/attestor.rs:345]: Move to above `sign_presentation` method in the `Holder` trait

Comment (edchapman88) [reply trustchain-ion/src/attestor.rs:345]: I've moved it to the implementation of the trait in tc-ion, where we introduce the principal of enforcing that the holder's field is populated

File: trustchain-ion/tests/vc.rs [hunk 12]
```diff
@@ -95,3 +95,10 @@ async fn test_sign_credential_failure() {
         Err(IssuerError::SSI(ssi::error::Error::KeyMismatch))
     ));
 }
+
+// TODO: add VP integration test
```
[the remark this answers is not in the document]
Comment (sgreenbury) [reply trustchain-ion/tests/vc.rs]: Now removed as duplicate of tests included in `trustchain-api/src/api.rs`

File: trustchain-api/src/api.rs [hunk 13]
```diff
@@ -119,5 +127,242 @@ pub trait TrustchainVCAPI {
...
+                        credential,
+                        ldp_options,
+                        root_event_time,
+                        verifier,
+                    )
+                    .await
+                    .map(|_| ())
+                    .map_err(|err| err.into()),
+                    CredentialOrJWT::JWT(jwt) => {
+                        // TODO: add chain verification for JWT credentials.
```
Comment (sgreenbury) [inline trustchain-api/src/api.rs:180]: I think we can use this [method](https://docs.rs/ssi/0.4.0/ssi/vc/struct.Credential.html#method.decode_verify_jwt) to extract the credential with a subsequent `TrustchainAPI::verify_credential` call?

File: trustchain-api/src/api.rs [hunk 14]
```diff
@@ -1,20 +1,28 @@
...
     vc::CredentialError,
     verifier::{Timestamp, Verifier, VerifierError},
+    vp::PresentationError,
 };
-use trustchain_ion::{attest::attest_operation, attestor::IONAttestor, create::create_operation};
+use trustchain_ion::{
+    attest::attest_operation, attestor::IONAttestor, create::create_operation, get_ion_resolver,
+    verifier::IONVerifier,
+};
+
```
Comment (sgreenbury) [inline trustchain-api/src/api.rs:24]:
```suggestion
```

File: trustchain-api/src/api.rs (hunk 13 above)
Comment (edchapman88) [reply trustchain-api/src/api.rs]: addressed with 24bdd3bb97b62197c7edf901b1998e2d8343ad8a

File: trustchain-api/src/api.rs [hunk 15]
```diff
@@ -119,5 +126,279 @@ pub trait TrustchainVCAPI {
...
+            .as_ref()
+            .ok_or(PresentationError::NoCredentialsPresent)?;
+
+        // TODO: consider concurrency limit (as rate limiting for verifier requests)
+        let limit = Some(5);
+        let ldp_options_vec: Vec<Option<LinkedDataProofOptions>> = (0..credentials.len())
+            .map(|_| ldp_options.clone())
+            .collect();
+
+        // Verify signatures and issuers for each credential included in the presentation
```
Comment (sgreenbury) [inline trustchain-api/src/api.rs:164]:
```suggestion
        // Verify signatures and issuers for each credential included in the presentation
        // TODO: consider concurrency limit (as rate limiting for verifier requests)
        let limit = Some(5);
        let ldp_options_vec: Vec<Option<LinkedDataProofOptions>> = (0..credentials.len())
            .map(|_| ldp_options.clone())
            .collect();
```

File: trustchain-core/src/vp.rs [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    #[error("No holder.")]
+    NoHolderPresent,
+    /// Wrapped variant for Trustchain holder.
+    #[error("A wrapped Trustchain holder error: {0}")]
+    HolderError(HolderError),
+    /// Wrapped variant for Crediential Error.
+    #[error("A wrapped Credential error: {0}")]
+    CredentialError(CredentialError),
+    /// Wrapped variant for Verifier Error.
+    #[error("A wrapped Verfier error: {0}")]
```
Comment (sgreenbury) [inline trustchain-core/src/vp.rs:22]:
```suggestion
    #[error("A wrapped Verifier error: {0}")]
```

File: trustchain-core/src/vp.rs [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    /// Wrapped variant for Verifier Error.
+    #[error("A wrapped Verifier error: {0}")]
+    VerifierError(VerifierError),
+    /// Credentials verified, but holder failed to authenticate with invalid or missing presentation
+    /// proof.
+    #[error("Credentials verified for an unauthenticated holder: {0:?}")]
+    VerifiedHolderUnauthenticated(VerificationResult),
+    /// Credentials verified, but holder DID failed verification (not part of valid Trustchain).
+    #[error("Credentials verified for an unverified holder: {0}")]
+    VerifiedHolderUnverfied(VerifierError),
```
Comment (sgreenbury) [inline trustchain-core/src/vp.rs:30]:
```suggestion
```

Remove as unused and potentially misleading variant

File: trustchain-ion/src/attestor.rs [hunk 18]
```diff
@@ -168,11 +169,59 @@ impl Issuer for IONAttestor {
...
+                if holder != &self.did {
+                    return Err(HolderError::MismatchedHolder);
+                }
+            }
+            None => vp.holder = Some(URI::String(self.did.clone())),
+        };
+
+        // Generate proof
+        // Example of verification_method derivation (optionally passed as a LinkedDataProofOption)
+        // let vm = format!("{}#{}", self.did(), signing_key.thumbprint().unwrap());
```
Comment (sgreenbury) [inline trustchain-ion/src/attestor.rs:211]:
```suggestion
```

**Review by sgreenbury — APPROVED** [review]
Looks good!

**Pull request merged**